File: Data Structures & Algorithms/top-k-elements-in-list/submission-5.py
class Solution:
    def topKFrequent(self, nums: List[int], k: int) -> List[int]:
        # Count the frequencies of each num in the nums list
        count = {}
        for num in nums:
            count[num] = 1 + count.get(num, 0)
        
        # Use a min-heap to pop the smallest frequency of each num
        # out of the heap
        heap = []
        for num in count.keys():
            heapq.heappush(heap, (count[num], num)) # pushes a tuple of (frequency, num)
            if len(heap) > k:
                heapq.heappop(heap)

        # Append k elements from the heap into the result array
        # Only use the num of the tuple (frequency, num)
        res = []
        for i in range(k):
            res.append(heapq.heappop(heap)[1])
        
        return res

File: Data Structures & Algorithms/products-of-array-discluding-self/submission-5.cpp
class Solution {
public:
    vector<int> productExceptSelf(vector<int>& nums) {
        int n = nums.size();
        vector<int> prefixes(n);
        vector<int> postfixes(n);
        vector<int> result(n);

        prefixes[0] = 1;
        postfixes[n - 1] = 1;

        for (int i = 1; i < n; i++) {
            prefixes[i] = nums[i - 1] * prefixes[i - 1];
        }

        for (int i = n - 2; i >= 0; i--) {
            postfixes[i] = nums[i + 1] * postfixes[i + 1];
        }

        for (int i = 0; i < n; i++) {
            result[i] = postfixes[i] * prefixes[i];
        }

        return result;
    }
};
